--- src/cairo-path-fixed.c.old	2013-11-24 20:49:54.000000000 +0100
+++ src/cairo-path-fixed.c	2013-11-24 21:03:39.000000000 +0100
@@ -1279,16 +1279,19 @@
  */
 cairo_bool_t
 _cairo_path_fixed_is_rectangle (const cairo_path_fixed_t *path,
-				cairo_box_t        *box)
+                                cairo_box_t        *box)
 {
     const cairo_path_buf_t *buf;
 
     if (! _cairo_path_fixed_is_box (path, box))
-	return FALSE;
+        return FALSE;
 
+    /* This check is valid because the current implementation of
+     * _cairo_path_fixed_is_box () only accepts rectangles like:
+     * move,line,line,line[,line|close[,close|move]]. */
     buf = cairo_path_head (path);
-    if (buf->points[0].y == buf->points[1].y)
-	return TRUE;
+    if (buf->num_ops > 4)
+        return TRUE;
 
     return FALSE;
 }
